---
title: "GSE164621_wt_mmu_lung"
author: "Mikias HW"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
---
# Prep Workspace
## Load libraries
```{r message=FALSE, warning=FALSE}
library(tidyverse)  # Data manipulation and visualization

library(Seurat)  # Single-cell RNA sequencing

library(patchwork)  # Combine plots

library(conflicted)  # Resolve function name conflicts
```

## Conflicted functions
```{r results='hide', message=FALSE, warning=FALSE}
# Set function preferences
conflicts_prefer(dplyr::select)

conflicts_prefer(dplyr::filter)

```

## Define Directories
```{r}
# Define data directories
dataDir <- paste(getwd(), "data", "GSE164621", "GSE164621_RAW", sep = "/")

dataFils <- list.files(dataDir)

datWT1 <- paste(dataDir, dataFils[1], sep = "/")
datWT2 <- paste(dataDir, dataFils[2], sep = "/")
datWT3 <- paste(dataDir, dataFils[3], sep = "/")

# Define output directory
outDir <- paste(getwd(), "output", "wtLung", sep = "/")
```

# Load Data
```{r}
# Load dataset
wt1Dat <- Read10X(datWT1)
wt2Dat <- Read10X(datWT2)
wt3Dat <- Read10X(datWT3)

# Initialize Seurat object 
wt1 <- CreateSeuratObject(counts = wt1Dat, project = "wt1", min.cells = 3, min.features = 200)
wt2 <- CreateSeuratObject(counts = wt2Dat, project = "wt2", min.cells = 3, min.features = 200)
wt3 <- CreateSeuratObject(counts = wt3Dat, project = "wt3", min.cells = 3, min.features = 200)
```


## Visualize QC
### violin plot
```{r fig.width=16, fig.height=10}
# Vizualize QC metrics and cutoffs
CountPlot <- ggplot(wt1@meta.data, aes("", nCount_RNA)) +
  geom_violin(fill = "pink") +
  geom_jitter(alpha = 0.2) 

  # geom_hline(yintercept = c(200, 7000), color = c("red")) +
  # labs(title = "Molecule Counts")

FeaturePlot <- ggplot(wt1@meta.data, aes("", nFeature_RNA)) +
  geom_violin(fill = "pink") + 
  geom_jitter(alpha = 0.2) 


  # geom_hline(yintercept = c(300, 1700), color = c("red", "red")) +
  # labs(title = "Unique Genes")

# Arrange plts
CountPlot + FeaturePlot + plot_layout(ncol = 2)

# Scatterplot
ggplot(wt1@meta.data, aes(nCount_RNA, nFeature_RNA)) +
  geom_point()


```

# Quickest Workflow
```{r eval=FALSE}
wt1 <- CreateSeuratObject(wt1Dat) %>%
    PercentageFeatureSet(pattern = "^MT-", col.name = "percent.mt") %>%
    SCTransform(vars.to.regress = "percent.mt") %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    RunUMAP(dims = 1:30) %>%
    FindClusters()

wt1
# Elbowplot
ElbowPlot(wt1)
```

# UMAP
```{r}
UMAPPlot(wt1, reduction = "umap") + labs(title="Clustered UMAP")
```

# Find Markers
```{r}
# find markers all cluster compared to all remaining cells, report only the positive ones
wt1.markers <- FindAllMarkers(wt1, only.pos = TRUE)
```

## Heatmap
```{r fig.height=16, fig.width=30}
top10 <- wt1.markers |> 
  group_by(cluster) |> 
  filter(avg_log2FC > 1) |> 
  slice_head(n = 10) |> 
  ungroup()

DoHeatmap(wt1, features = top10$gene) + NoLegend()
```

## Widefield View
```{r fig.height=50, fig.width=16, warning=FALSE, message=FALSE}
# Extract top 3 cluster identifying markers
ClusID <- top10 |> 
  select(cluster, gene) |>
  group_by(cluster) |> 
  slice_head(n=3) |> 
  ungroup()

# UMAP visualization 
FeaturePlot(wt1, features = ClusID$gene, ncol = 3)

# Violin plots
VlnPlot(wt1, features = ClusID$gene, ncol = 3)

# Ridgeplots
RidgePlot(wt1, features = ClusID$gene, ncol = 3)

# Dotplot
DotPlot(wt1, features =  ClusID$gene)
```